client/service/tofu.go

```diff
@@ -0,0 +1,184 @@
+package boxservice
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"time"
+)
+
+type CertDetail struct {
+	Subject     string
+	Issuer      string
+	Fingerprint string
+}
+
+// this file contains the code for Trust On First Use (TOFU) when accessing lantern server manager
+var (
+	ErrNoMatchingFingerprint = errors.New("server certificate(s) didn't match trusted fingerprint")
+	ErrCertExpired           = errors.New("server certificate has expired")
+	ErrCertNotYetValid       = errors.New("server certificate is not yet valid")
+	ErrNoCertsDetected       = errors.New("no server certificates detected")
+	ErrTrustCancelled        = errors.New("user cancelled the trust dialog")
+)
+
+// getFingerprint returns the SHA1 fingerprint of the given DER-encoded certificate.
+func getFingerprint(der []byte) string {
+	hash := sha1.Sum(der)
+	hexified := make([][]byte, len(hash))
+	for i, data := range hash {
+		hexified[i] = []byte(fmt.Sprintf("%02X", data))
+	}
+	return string(bytes.Join(hexified, []byte(":")))
+}
+
+// getTOFUClient return an http client that will verify a match given a trusted fingerprint,
+func getTOFUClient(fingerprint string) (*http.Client, error) {
+	slog.Debug("Attempting connection with trusted fingerprint", "fingerprint", fingerprint)
+	dial := func(_ context.Context, network, addr string) (net.Conn, error) {
+		config := &tls.Config{
+			InsecureSkipVerify: true, // We'll verify ourselves
+		}
+
+		conn, err := tls.Dial(network, addr, config)
+		if err != nil {
+			slog.Error("Unable to connect", "addr", addr, "err", err)
+			return nil, err
+
+		}
+		state := conn.ConnectionState()
+		now := time.Now()
+		matched := false
+		for _, cert := range state.PeerCertificates {
+			if fingerprint == getFingerprint(cert.Raw) {
+				matched = true
+			}
+			if now.Before(cert.NotBefore) {
+				_ = conn.Close()
+				return nil, ErrCertNotYetValid
+			}
+			if now.After(cert.NotAfter) {
+				_ = conn.Close()
+				return nil, ErrCertExpired
+			}
+		}
+		if !matched {
+			_ = conn.Close()
+			return nil, ErrNoMatchingFingerprint
+		}
+		// If we've gotten this far, then we can trust the server
+		slog.Debug("Server cert(s) passed TOFU tests")
+		return conn, nil
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			DialTLSContext: dial,
+		},
+	}, nil
+}
+
+// getServerFingerprints connects to an untrusted server, and get cert information about it
+// the addr must be a valid address, including port
+func getServerFingerprints(addr string) ([]CertDetail, error) {
+	var ret []CertDetail
+
+	// If the port isn't specified, assume https standard port
+	if !strings.Contains(addr, ":") {
+		addr = addr + ":443"
+	}
```

Comment:
This will trigger a false positive for bare IPv6 addresses. Probably better to use [`net.SplitHostPort()`](https://pkg.go.dev/net#SplitHostPort) and check for an empty `port` return value.